abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+# Proxy-Wasm ABI v0.2.1 specification
+
+---
+
+Status: **FINAL**
+
+---
+
+# Callbacks and Functions
+
+### Callbacks exposed by the Wasm module
+
+All callbacks (entry points named `proxy_on_<event>`) are optional,
+and are going to be called only if exposed by the Wasm module.
+
+All callbacks, other than the integration functions, include a unique
+context identifier as the first parameter, which can be used to
+distinguish between different contexts.
+
+
+### Functions exposed by the host environment
+
+All functions exposed by the host return [`proxy_status_t`], which
+indicates status of the call (success, invalid memory access, etc.),
+and return values are written into memory pointed by `return_<value>`
+parameters.
+
+
+## Integration
+
+### `proxy_abi_version_0_2_1`
+
+* params:
+  - none
+* returns:
+  - none
+
+Function marker used during linking to advertise support for
+Proxy-Wasm ABI v0.2.1.
+
+This function is never called.
+
+
+### `_initialize`
+
+* params:
+  - none
+* returns:
+  - none
+
+Called when the Wasm module is first loaded.
+
+
+### `main`
+
+* params:
+  - `i32 (uint32_t) ignored`
+  - `i32 (uint32_t) ignored`
+* returns:
+  - `i32 (uint32_t) ignored`
+
+> **Warning**
+> This is called only if [`_initialize`] is also exported.
+
+Called when the Wasm module is first loaded, after [`_initialize`].
+
+
+### `_start`
+
+* params:
+  - none
+* returns:
+  - none
+
+> **Warning**
+> This is called only if [`_initialize`] is not exported.
+
+Called when the Wasm module is first loaded.
+
+
+### `proxy_on_memory_allocate`
```

Comment:
maybe somewhere say that either `proxy_on_memory_allocate` or `malloc` may be absent (not exported by the guest), but not both?

Reply:
This is a little tricky. One of those callbacks is required in the current Envoy implementation, but technically plugins don't need to export this if they don't retrieve strings from the host.